### PR TITLE
Add spec conformance tests for nil-literal

### PR DESCRIPTION
**Pull request by SandaruJayawardana**
Related lang issue -[ #32456](https://github.com/ballerina-platform/ballerina-lang/issues/32456)

### CODE REVIEW

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+Test-Case: output
+Description: Test nil type descriptor with nil literal 'null'.
+Labels: nil-type-descriptor, nil-literal
+
+public function main() {
+    () x = null;
+    io:println(x.toBalString()); // @output ()
+}
+
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/nil-literal.balt:18]: Shall we duplicate these two test cases, without using toBalString(). So the output will be just an empty string. This is because there are no test cases covering normal toString behavior.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 2]
```diff
@@ -0,0 +1,173 @@
+Test-Case: output
+Description: Test nil type descriptor with nil literal '()'.
+Labels: nil-type-descriptor, nil-literal
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/nil-literal.balt:3]: Include `lang.value` label, in all the places where we use toBalString()

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+Description: Test nil type descriptor with nil literal 'null'.
+Labels: nil-type-descriptor, nil-literal
+
+public function main() {
+    () x = null;
+    io:println(x.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test value and assignability of the nil literal.
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/nil-literal.balt:20]: Should be equality, Word `assignability` is not used in the spec.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+Labels: nil-type-descriptor, nil-literal, equality-expr
+
+public function main() {
+    () x = ();
+    () y = null;
+    io:println(x == y); // @output true
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/nil-literal.balt:26]: This is part of the nil type, Since are testing the nil literal here, we can write this as `() == null`. Also we can include `!=` `===` and `!==` with nil literal. `is` expression another place we can use nil literal. Please add a test case to cover that.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 5]
```diff
@@ -0,0 +1,173 @@
+Test-Case: output
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/nil-literal.balt:1]: Should be part of the `expressions` not `values-types-and-variables`

Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/nil-literal.balt:1]: will correct it.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    () x = ();
+    () y = null;
+    io:println(x == y); // @output true
+
+    x = y;
+    io:println(x.toBalString()); // @output ()
+
+    x = ();
+    y = x;
+    io:println(y.toBalString()); // @output ()
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/nil-literal.balt:33]: These aren't quite relevant to nil literal?

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+    x = (
+    );
+    io:println(x.toBalString()); // @output ()
+
+    x = (());
+    io:println(x.toBalString()); // @output ()
+
+    x = (null);
+    io:println(x.toBalString()); // @output ()
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/nil-literal.balt:55]: Let's make just this a separate test.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    float? c = null;
+    io:println(c.toBalString()); // @output ()
+
+    int? d = null;
+    io:println(d.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test nil literals with return statement.
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/nil-literal.balt:121]: IMO, the more accurate test for this is `return ();` right?

Comment (MaryamZi) [reply conformance/lang/values-types-and-variables/nil-literal.balt:121]: Let's add one for null also.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    float? c = ();
+    io:println(c.toBalString()); // @output ()
+
+    int? d = ();
+    io:println(d.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test nil literal 'null' with optional type descriptor.
+Labels: boolean-type-descriptor, floating-point-type-descriptor, int-type-descriptor, optional-type-descriptor,
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/nil-literal.balt:103]: Didn't we say these would come under optional type-descriptor tests?

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+Labels: nil-type-descriptor
+
+public function main() {
+    () a = Null; // @error expected a '()', but found an undefined symbol 'Null'
+
+    () b = NULL; // @error expected a '()', but found an undefined symbol 'NULL'
+
+    () c = nil; // @error expected a '()', but found an undefined symbol 'nil'
+
+    () d = Nil; // @error expected a '()', but found an undefined symbol 'Nil'
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/nil-literal.balt:172]: We can maybe add few more simple tests like `() x = ());`.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+    x = <decimal> (); // @error '()' cannot be cast to 'decimal'
+
+    x = <float> (); // @error '()' cannot be cast to 'float'
+
+    x = <int> (); // @error '()' cannot be cast to 'int'
+
+    x = <string> (); // @error '()' cannot be cast to 'string'
+}
+
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/nil-literal.balt:160]: We can write a few more tests.

For example
- Method call on ()/null (`().toString()`, `().toBalString()`)
- ()/null as interpolation in template expressions
- ()/null as member of structural constructor

Let's structure the tests as expressions first and then actions/statements.

File: conformance/lang/values-types-and-variables/nil-literal.balt [hunk 12]
```diff
@@ -0,0 +1,173 @@
+Test-Case: output
+Description: Test nil type descriptor with nil literal '()'.
```
[the remark this answers is not in the document]
Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/nil-literal.balt:2]: Okay. will update.

File: conformance/lang/values-types-and-variables/nil-literal.balt (hunk 6 above)
Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/nil-literal.balt:33]: Removed them.

File: conformance/lang/values-types-and-variables/nil-literal.balt (hunk 8 above)
Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/nil-literal.balt:121]: Yes. will update.

File: conformance/lang/values-types-and-variables/nil-literal.balt (hunk 9 above)
Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/nil-literal.balt:103]: I think we had decided to add few test cases related to this area.

File: conformance/lang/expressions/literals/nil_literal.balt [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+    () c = nul; // @error expected a '()', but found an undefined symbol 'nul'
+
+    () d = nil; // @error expected a '()', but found an undefined symbol 'nil'
+
+    () e = Nil; // @error expected a '()', but found an undefined symbol 'Nil'
+
+    () f = ()); // @error missing '(' token
+
+    () g = ); // @error missing '(' token
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/nil_literal.balt:260]: There are parser tests, right? Shall we move them out to a separate test case?

**Review by MaryamZi — COMMENTED** [review]
@SandaruJayawardana, 4 tests seems to be failing for me. Can you validate with the test runner PR?

File: conformance/lang/expressions/literals/nil_literal.balt [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+
+public function main() {
+    io:println(testNil().toBalString()); // @output ()
+}
+
+function testNil() returns () {
+    return null;
+}
+
+Test-Case: output
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/nil_literal.balt:145]: Shall we also add simple JSON tests with both nil and null, because the section on the nil literal specifically mentions JSON and using null with JSON?

Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/nil_literal.balt:145]: will add.

File: conformance/lang/expressions/literals/nil_literal.balt (hunk 13 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/nil_literal.balt:260]: Yes. Will change.

File: conformance/lang/expressions/literals/nil_literal.balt [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+Description: Test nil literal '()' with function call.
+Labels: function-call-expr, nil-literal, nil-type, value:toBalString
+
+public function main() {
+    testNil(());
+}
+
+function testNil(() x) {
+    io:println(x.toBalString()); // @output ()
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/nil_literal.balt:128]: Output should be in L124?

Comment (MaryamZi) [reply conformance/lang/expressions/literals/nil_literal.balt:128]: Please update wherever applicable.

File: conformance/lang/expressions/literals/nil_literal.balt [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+    float? c = null;
+    io:println(c.toBalString()); // @output ()
+
+    int? d = null;
+    io:println(d.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test nil literal 'null' with json type descriptor.
+Labels: null, json-type-descriptor, value:toBalString
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/nil_literal.balt:267]: Let's use `json` as the label.

File: conformance/lang/expressions/literals/nil_literal.balt (hunk 13 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/nil_literal.balt:300]: Sure. Didn't notice it.

**Review by hasithaa — APPROVED** [review]
